pkg/rbd/controllerserver.go

```diff
@@ -150,6 +150,14 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	if found {
+		if rbdVol.Encrypted {
+			err = ensureEncryptionMetadataSet(ctx, cr, rbdVol)
```

Comment:
If this fails, the CO has no way to garbage collect the volume created thus far (unlike calling `deleteImage` in line 193 below). The reason being the vol reservation is cancelled only if we fail to find the image, and that will not happen as the image is never deleted, so the next create volume request will attempt to use the same image and reset the key.

The solution I think falls into moving the encryption key to the RADOS OMap based journal, rather than retaining it in the image itself (I understand this is contradictory to earlier review, but I think this is better than amending image meta attributes post creating the image).

If we retain the encryption status on the CSI image RADOS OMap, we need not bother with attempting to set it post the image is created as well, which is better when returning found status for an existing image, than modifying it post it being found.

As this is data that is stored on disk (IOW, the image meta or the RADOS key), we need to evaluate this choice before we merge this change.

Reply:
my takeaway on this would be that using image-meta is better since this is closer to the image itself which means safer solution as we don't have a separate object somewhere holding crucial information...

if I'm getting it right, the next request will try to use the same image, so it will reset the encryption key on this image and will attempt second time image-meta. And this becomes an issue only if this call fails persistently and not a once-off failure due to network issues for example.

so personally I would prefer to be on a safer side to have information about the image on the image itself, but happy to hear more suggestions

Reply:
The workflow I would like to see for `CreateVolume` when returning an existing volume, is one where it does not modify any volume attributes when it finds a volume matching the request.

This is important if and when we support `ListVolumes`, as in that case we cannot change (set/reset) volume attributes for all encrypted volumes that we list, as some may be already in use by then.

The [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume-deletevolume-listvolumes) calls out that, the CO can send repeated `CreateVolume` or `ListVolumes` to rebuild its state, or to retry and get a `volume_id` for a timed out request. Which makes the above (`ListVolumes`) possible in the future.

For the above reasons, and also as an implementation workflow, I want to ensure there is no modification of any attribute of a volume, when we determine a pre-existing volume matching the criteria in a `CreateVolume` request.

Coming to where we maintain the encrypted flag,

1. if we choose to maintain it on the image (for reasons stated in [this](https://github.com/ceph/ceph-csi/pull/697#discussion_r355543759) comment), then I would state we should delete the image and undo the reservation in the function [`checkVolExists`](https://github.com/ceph/ceph-csi/blob/master/pkg/rbd/rbd_journal.go#L159). This will recreate the requested volume and ensure we finish all steps (including setting up the image key) before returning a success as a part of a single running transaction.
Even with the above, when we start supporting `ListVolumes`, we would return a volume that has a `volume_context` which would state it is not encrypted if the image key was not set (due to interrupted transaction on a prior `CreateVolume` request), but the reservation and the image were found. Also, we now need to handle image delete as part of the `undoVolReservation`, or just before the same.

1. If we maintain the encrypted flag, as part of the CSI OMap for the image, we would have created the image post setting up the required attributes (in this case in the CSI OMap), and the workflow in `checkVolExists` can flow as it is currently, with an additional check for the added encrypted key. There would be no image deletion that would need to be done and `ListVolumes` can respond with concrete data as the reservation, its attributes and the image is found on disk.

I hence prefer that we store the encrypted (and other future keys) in the CSI OMap for the image, rather than the image itself.

The further workflow ensures that `NodeStageVolume` gets the `volume_context` that enables working with encrypted volumes without checking the actual OMap or the image meta. `NodeStageVolume` though will have to update either the OMap to reflect that the volume is already encrypted, or we can add a image key, when we actually encrypt the volume, to state it is encrypted.

IOW, `NodeStageVolume` can use image keys, and `CreateVolume` can use the CSI OMaps to store the encryption status. (This is assuming we can tackle not re-encrypting an image as discussed [here](https://github.com/ceph/ceph-csi/pull/697#discussion_r355185194))

Reply:
Based on a discussion with @dillaman it is better to maintain the the encrypted key with the image, IOW as `image-meta` key:value. This helps copying the key across image clones, and hence preserving encryption status for clones as well.

Further, to prevent incomplete `CreateVolume` we will maintain an "in-progress" key in the CSI OMap, such that repeat `CreateVolumes` can choose to complete the transaction or can garbage collect the image and the reservation and start afresh. This is not a necessary concern for this patch, IOW can be added subsequently as well.

This issue will hence be marked resolved once we create an issue to track the above. Also, this comment no longer blocks this PR.

Reply:
Suggested improvements tracked in #744